kubernetes/check.py

```diff
@@ -399,46 +425,37 @@ def _update_node(self, instance):
         # reserved for system/Kubernetes.
 
     def _update_pods_metrics(self, instance, pods):
-        supported_kinds = [
-            "DaemonSet",
-            "Deployment",
-            "Job",
-            "ReplicationController",
-            "ReplicaSet",
-        ]
-
-        # (create-by, namespace): count
-        controllers_map = defaultdict(int)
-        for pod in pods['items']:
-            try:
-                created_by = json.loads(pod['metadata']['annotations']['kubernetes.io/created-by'])
-                kind = created_by['reference']['kind']
-                if kind in supported_kinds:
-                    namespace = created_by['reference']['namespace']
-                    controllers_map[(created_by['reference']['name'], namespace)] += 1
-            except (KeyError, ValueError) as e:
-                self.log.debug("Unable to retrieve pod kind for pod %s: %s", pod, e)
-                continue
-
-        tags = instance.get('tags', [])
-        for (ctrl, namespace), pod_count in controllers_map.iteritems():
-            _tags = tags[:]  # copy base tags
-            _tags.append('kube_replication_controller:{0}'.format(ctrl))
-            _tags.append('kube_namespace:{0}'.format(namespace))
-            self.publish_gauge(self, NAMESPACE + '.pods.running', pod_count, _tags)
-
-    def _process_events(self, instance, pods_list):
         """
-        Retrieve a list of events from the kubernetes API.
+        Reports the number of running pods, tagged by service and creator
 
-        At the moment (k8s v1.3) there is no support to select events based on a timestamp query, so we
-        go through the whole list every time. This should be fine for now as events
-        have a TTL of one hour[1] but logic needs to improve as soon as they provide
-        query capabilities or at least pagination, see [2][3].
-
-        [1] https://github.com/kubernetes/kubernetes/blob/release-1.3.0/cmd/kube-apiserver/app/options/options.go#L51
-        [2] https://github.com/kubernetes/kubernetes/issues/4432
-        [3] https://github.com/kubernetes/kubernetes/issues/1362
+        We go though all the pods, extract tags then count them by tag list, sorted and
+        serialized in a pipe-separated string (it is an illegar character for tags)
+        """
+        tags_map = defaultdict(int)
+        for pod in pods['items']:
+            pod_meta = pod.get('metadata', {})
+            pod_tags = self.kubeutil.get_pod_creator_tags(pod_meta, legacy_rep_controller_tag=True)
+            services = self.kubeutil.match_services_for_pod(pod_meta)
+            if isinstance(services, list):
+                for service in services:
+                    pod_tags.append('kube_service:%s' % service)
+            if 'namespace' in pod_meta:
+                pod_tags.append('kube_namespace:%s' % pod_meta['namespace'])
+
+            pod_tags.sort()
+            tags_map[TAG_SEPARATOR.join(pod_tags)] += 1
```

Comment:
set is mutable too, so it can't be used as a dict key
But I could use a frozenset https://docs.python.org/2/library/stdtypes.html#set-types-set-frozenset instead